### PR TITLE
Upgrading Terraform to 0.12.3

### PR DESCRIPTION
**Pull request by aLekSer**
Fixed configuration tf files according to a new format.
`terraform fmt` all files.
Also updated providers versions.

As part of #657

### CODE REVIEW

File: build/cluster.tf [hunk 1]
```diff
@@ -67,15 +73,15 @@ locals {
...
 resource "google_container_cluster" "primary" {
-  name     = "${lookup(var.cluster, "name")}"
-  location     = "${lookup(var.cluster, "zone")}"
-  project  = "${lookup(var.cluster, "project")}"
-  provider = "google-beta"
+  name               = "${var.cluster["name"]}"
+  location           = "${var.cluster["zone"]}"
+  project            = "${var.cluster["project"]}"
+  provider           = "google-beta"
+  enable_legacy_abac = "${var.cluster["legacyAbac"]}"
```
Comment (roberthbailey) [inline build/cluster.tf]: Please remove this. legacy ABAC has been disabled for a long time on GKE and I'm removing this option from the various ways to create GKE clusters.

File: build/cluster.tf [hunk 2]
```diff
@@ -67,15 +73,15 @@ locals {
...
 
   # Setting an empty username and password explicitly disables basic auth
```
Comment (roberthbailey) [inline build/cluster.tf]: Please remove the master_auth block. Now that Mark switched GKE to 1.12 the defaults in GKE disable both basic auth and client certs.

File: build/cluster.tf [hunk 3]
```diff
@@ -87,86 +93,94 @@ resource "google_container_cluster" "primary" {
...
+      tags = ["game-server"]
+    }
+  }
+
+  node_pool {
+    name       = "agones-system"
+    node_count = 1
+
+    node_config {
+      preemptible  = true
```
Comment (roberthbailey) [inline build/cluster.tf]: Why are we putting the agones controller onto a preemptible vm?

File: build/includes/terraform.mk [hunk 4]
```diff
@@ -39,7 +39,7 @@ endif
 	 	-var "cluster={name=\"$(GCP_CLUSTER_NAME)\", machineType=\"$(GCP_CLUSTER_NODEPOOL_MACHINETYPE)\", \
 		 zone=\"$(GCP_CLUSTER_ZONE)\", project=\"$(GCP_PROJECT)\", \
 		 initialNodeCount=\"$(GCP_CLUSTER_NODEPOOL_INITIALNODECOUNT)\", \
-		 legacyABAC=\"$(GCP_CLUSTER_LEGACYABAC)\"}"'
+		 legacyAbac=\"$(GCP_CLUSTER_LEGACYABAC)\"}"'
```
Comment (roberthbailey) [inline build/includes/terraform.mk]: Please remove legacyAbac

File: build/includes/terraform.mk [hunk 5]
```diff
@@ -65,7 +65,7 @@ endif
 		-var "cluster={name=\"$(GCP_CLUSTER_NAME)\", machineType=\"$(GCP_CLUSTER_NODEPOOL_MACHINETYPE)\", \
 		 zone=\"$(GCP_CLUSTER_ZONE)\", project=\"$(GCP_PROJECT)\", \
 		 initialNodeCount=\"$(GCP_CLUSTER_NODEPOOL_INITIALNODECOUNT)\", \
-		 legacyABAC=\"$(GCP_CLUSTER_LEGACYABAC)\"}"'
+		 legacyAbac=\"$(GCP_CLUSTER_LEGACYABAC)\"}"'
```
Comment (roberthbailey) [inline build/includes/terraform.mk]: Please remove legacyAbac

File: build/modules/gke/cluster.tf [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
 provider "google-beta" {
   version = "~> 2.4"
-  zone      = "${lookup(var.cluster, "zone")}"
+  zone    = "${lookup(var.cluster, "zone")}"
```
Comment (roberthbailey) [inline build/modules/gke/cluster.tf]: In the other file you changed places that had previously done a `lookup(...)` to a different syntax (`${var.cluster["zone"]}`) but this file still uses the `lookup(...)` syntax. Can you make this file consistent with `build/cluster.tf`?

File: build/modules/gke/cluster.tf [hunk 7]
```diff
@@ -14,7 +14,7 @@
 
 provider "google-beta" {
   version = "~> 2.4"
```
Comment (roberthbailey) [inline build/modules/gke/cluster.tf]: 2.10 (like above)?

File: build/modules/gke/cluster.tf [hunk 8]
```diff
@@ -28,15 +28,14 @@ data "google_client_config" "default" {}
 # echo command used for debugging purpose
 # Run `terraform taint null_resource.test-setting-variables` before second execution
 resource "null_resource" "test-setting-variables" {
-    provisioner "local-exec" {
-      command = "${"${format("echo Current variables set as following - name: %s, project: %s, machineType: %s, initialNodeCount: %s, zone: %s, legacyAbac: %s",
+  provisioner "local-exec" {
+    command = "${"${format("echo Current variables set as following - name: %s, project: %s, machineType: %s, initialNodeCount: %s, zone: %s, legacyAbac: %s",
```
Comment (roberthbailey) [inline build/modules/gke/cluster.tf]: Please remove legacyAbac here too.

File: build/modules/gke/cluster.tf [hunk 9]
```diff
@@ -45,15 +44,16 @@ locals {
...
 }
 
 resource "google_container_cluster" "primary" {
   name     = "${lookup(var.cluster, "name")}"
-  location     = "${lookup(var.cluster, "zone")}"
+  location = "${lookup(var.cluster, "zone")}"
   project  = "${lookup(var.cluster, "project")}"
   provider = "google-beta"
+
   # Setting an empty username and password explicitly disables basic auth
```
Comment (roberthbailey) [inline build/modules/gke/cluster.tf]: Please remove the whole master_auth block now that we are using 1.12.

File: build/cluster.tf (hunk 2 above)
Comment (aLekSer) [reply build/cluster.tf]: Thanks for the information, will update soon

File: build/cluster.tf (hunk 3 above)
Comment (aLekSer) [reply build/cluster.tf]: Nice catch, I think this is a left-over from other Terraform configuration example.

File: build/modules/gke/cluster.tf (hunk 6 above)
Comment (aLekSer) [reply build/modules/gke/cluster.tf]: Yes, for sure

**Comment by aLekSer** [comment]
While deploying Agones with Terraform I got sometimes timeout errors with `agones-ping-udp-service`. Screenshot of tiller-deployment is attached. 
Screenshot 2019-07-17 at 15 24 27
Some terraform deployment completed successfully.

**Comment by markmandel** [comment]
> While deploying Agones with Terraform I got sometimes timeout errors with `agones-ping-udp-service`. 

So does that mean that this PR is currently not stable?

**Comment by aLekSer** [comment]
Yes, this version is unstable, I am fixing this issues now.

File: build/cluster.tf [hunk 10]
```diff
@@ -13,160 +13,138 @@
...
+        "https://www.googleapis.com/auth/devstorage.read_only",
+        "https://www.googleapis.com/auth/logging.write",
+        "https://www.googleapis.com/auth/monitoring",
+        "https://www.googleapis.com/auth/service.management.readonly",
+        "https://www.googleapis.com/auth/servicecontrol",
+        "https://www.googleapis.com/auth/trace.append",
+      ]
+
+      labels = {
+        "stable.agones.dev/agones-system" = "true"
```
Comment (roberthbailey) [inline build/cluster.tf]: The label (and taint) should just be `agones.dev/agones-system` now.

File: build/cluster.tf [hunk 11]
```diff
@@ -13,160 +13,138 @@
...
+        "https://www.googleapis.com/auth/devstorage.read_only",
+        "https://www.googleapis.com/auth/logging.write",
+        "https://www.googleapis.com/auth/monitoring",
+        "https://www.googleapis.com/auth/service.management.readonly",
+        "https://www.googleapis.com/auth/servicecontrol",
+        "https://www.googleapis.com/auth/trace.append",
+      ]
+
+      labels = {
+        "stable.agones.dev/agones-metrics" = "true"
```
Comment (roberthbailey) [inline build/cluster.tf]: The label and taint should be `agones.dev/agones-metrics`.

File: build/modules/gke/cluster.tf [hunk 12]
```diff
@@ -28,136 +28,70 @@ data "google_client_config" "default" {}
...
+        "https://www.googleapis.com/auth/devstorage.read_only",
+        "https://www.googleapis.com/auth/logging.write",
+        "https://www.googleapis.com/auth/monitoring",
+        "https://www.googleapis.com/auth/service.management.readonly",
+        "https://www.googleapis.com/auth/servicecontrol",
+        "https://www.googleapis.com/auth/trace.append",
+      ]
+
+      labels = {
+        "stable.agones.dev/agones-system" = "true"
```
Comment (roberthbailey) [inline build/modules/gke/cluster.tf]: as above, please fix the label / taint names.

File: build/modules/gke/variables.tf [hunk 13]
```diff
@@ -15,24 +15,31 @@
 # Password for the Kubernetes API.
 # Could be defined using GKE_PASSWORD env variable
 # or by setting `password="somepass"` string in build/terraform.tfvars
-variable "password" {default = ""}
-variable "username" {default = "admin"}
+variable "password" {
```
Comment (roberthbailey) [inline build/modules/gke/variables.tf]: It should be possible to remove the password and username variables.

File: build/modules/gke/cluster.tf [hunk 14]
```diff
@@ -28,136 +28,70 @@ data "google_client_config" "default" {}
...
+      }
+
+      taint {
+        key    = "stable.agones.dev/agones-system"
+        value  = "true"
         effect = "NO_EXECUTE"
+      }
```
Comment (roberthbailey) [inline build/modules/gke/cluster.tf]: This seems to have lost the node pool for agones-metrics.

File: build/cluster.tf (hunk 10 above)
Comment (aLekSer) [reply build/cluster.tf]: thanks for an update

File: build/modules/gke/cluster.tf (hunk 14 above)
Comment (aLekSer) [reply build/modules/gke/cluster.tf]: that's right

Comment (roberthbailey) [reply build/modules/gke/cluster.tf]: Was that on purpose or should the 3rd node pool be defined here?

Comment (aLekSer) [reply build/modules/gke/cluster.tf]: It should be there

**Comment by aLekSer** [comment]
Error in E2E:
```
Step #17: time="2019-07-23 13:15:26.656" level=info msg="fleet simple-fleet-2mhnt has 0/1 ready replicas"
Step #17: --- FAIL: TestGameServerUnhealthyAfterDeletingPod (86.77s)
Step #17: gameserver_test.go:211: 
Step #17: Error Trace:	gameserver_test.go:211
Step #17: Error: Received unexpected error:
Step #17: timed out waiting for the condition
Step #17: waiting for GameServer to be Unhealthy default/udp-serverk4lj7
Step #17: agones.dev/agones/test/e2e/framework.(*Framework).WaitForGameServerState
Step #17: /go/src/agones.dev/agones/test/e2e/framework/framework.go:127
Step #17: agones.dev/agones/test/e2e.TestGameServerUnhealthyAfterDeletingPod
Step #17: /go/src/agones.dev/agones/test/e2e/gameserver_test.go:210
Step #17: testing.tRunner
```

**Comment by aLekSer** [comment]
@roberthbailey  applied all your comments. Now scripts are working.
- `make gcloud-terraform-install` would install current version of agones, 
- `make gcloud-terraform-cluster` will install release version from agones helm chart.

**Pull request merged**